Fixes issue #1163

mikkpokk opened the pull request:
More info: https://github.com/dromru/react-photoswipe-gallery/issues/1163

mikkpokk committed:
Fixes issue #1163

More info: https://github.com/dromru/react-photoswipe-gallery/issues/1163

mikkpokk committed:
Replace ref with getRef

DanielRiveraHQ commented:
Thanks @mikkpokk, hopefully we can get this merged soon

MaximKondratev commented on src/item.ts:
```diff
@@ -94,19 +101,28 @@ export interface ItemProps {
+    }
+  }, [])
+
+  return children({ ref, open, getRef })
```
do we really need to expose any additional api? is there any reason for it? if not i would prefer to keep public api as simple.

mikkpokk replied:
Currently `react-photoswipe-gallery` generates ref from image itself, instead of passing in own ref (`forwardRef` technique). In such cases, sometimes developers need to access that very same ref in their custom component as well (ex. tracking viewport changes).

`useCallback` is not returning any object. While using `useCallback` for generating node ref, it can store ref only to another ref.

So yes, there is a reason. You cannot generate another node ref once you already done it. It is elegant to pass it back to developer.

`useLayoutEffect` has a pitfall and always useEffect should be prefered. I believe `useEffect` was replaced with `useLayoutEffect` because `useLayoutEffect` seemed to fix some kind of null node unhandled exception. However, that's not the case and not a one developer should use hook `useLayoutEffect` for such a cases. `useLayoutEffect` will be called right after DOM is loaded. However, it doesn't mean `react-photoswipe-gallery` is loaded at that point yet. `useLayoutEffect` can hurt performance and it will be never ran in SSR frameworks. Because SSR is not browser which can track when DOM is loaded to browser. So that's why I choosed `useCallback` method instead reverting back `useLayoutEffect` to `useEffect`.

MaximKondratev replied:
`useCallback` is a good decision here.  Moreover, around half year ago we planned to improve typing of `ref` using `useCallback` instead of raw `ref`. But didn't have time.

Yet I am still not sure about **new third argument** of `children`. The reason is I don't think we should provide a getter at all.

If you need that ref somewhere else you can use ref callback to save it as you mentioned in your comment https://github.com/dromru/react-photoswipe-gallery/pull/1212#issuecomment-1642931630

MaximKondratev replied:
I even found my work around ref callback: https://github.com/dromru/react-photoswipe-gallery/compare/master...MaximKondratev:react-photoswipe-gallery:change-ref-to-callback-ref 

It was made for typings first, but it also solves ssr problem.

so I propose we come back to my fork and finish it @irudoy? 

if you need fast fix we could use `useIsomorphicLayoutEffect` as a hack patch.

mikkpokk replied:
I checked your PR. It kind of changes too much things. For an example, you destroying current gallery functionality because you don't want to pass in undefined refs to Context. And I'm not sure it's easier to finish or fix your PR than merge already working solution right here.

In case you have plan to change typings, it's easier to countinue with the current PR.

As I stated before, `useLayoutEffect` is not causing only SSR warnings. It causes performance issues. So using `useIsomorphicLayoutEffect` instead of ` useLayoutEffect` is not even a patch.

MaximKondratev replied:
As both @irudoy and @akhmadullin are on vacation for one or two weeks I think they will not reply here. 

Generally what i don't like is adding new api to `children` (`getRef`) which is at least minor release and yet i'm not sure if it can be useful in real world scenario.

For now i also don't think i'm back at my pr soon (my work, others vacations and then my vacation) so we probably won't close it in days.

If you have that much performance issues for now you can try do downgrade version to 2.2.6.

mikkpokk replied:
There isn't new api to children. Inside `Item` component there is `getRef` to clean up photoswipe context. You cannot make clean ups inside callback. To use callback node outside of the callback it has to be passed to parent ref object.

Callback will be called only when children element or callback dependencies change.

MaximKondratev commented on src/item.ts:
```diff
@@ -16,7 +16,14 @@ interface ChildrenFnProps {
    * Can be omitted if there is only one item in gallery
    */
-  ref: ItemRef
+  ref: (node) => void
```
Not sure if we have typescript pipeline check. Looks like no type provided here and in useCallback

mikkpokk committed:
node: HTMLElement | null

mikkpokk committed:
fix: useLayoutEffect replaced with useCallback

Related with #1163

Keep in mind, usage actually stays the same. You can call `ref={ref}` as before.

mikkpokk commented:
Usage stays the same. You can use it as:

```jsx
<Gallery>
    <Item key={key}
              original={item.original}
              thumbnail={item.thumbnail}
              width={item.width}
              height={item.height}>
            {({ref, open}) => {
                return (
                    <img src={item.thumbnail} ref={ref} />
                )
            }}
    </Item>
</Gallery>
```

OR

```jsx
<Gallery>
    <Item key={key}
              original={item.original}
              thumbnail={item.thumbnail}
              width={item.width}
              height={item.height}>
            {({ref, open}) => {
                return (
                    <img src={item.thumbnail} ref={(node: HTMLImageElement) => {
                        // localCallback(node)
                        ref(node)
                    }} />
                )
            }}
    </Item>
</Gallery>
```

mikkpokk commented:
pages/deploy check failed due to permissions error. Please review.

mikkpokk commented:
@irudoy @akhmadullin @MaximKondratev

mikkpokk committed:
fix: useLayoutEffect replaced with useCallback

mikkpokk committed:
fix: typing

mikkpokk committed:
fix: typings
